tests/DataValues/TimeValueCalculatorTest.php

```diff
@@ -209,4 +209,153 @@ public function testGetSecondsForPrecision( $precision, $expected ) {
+		$timestamps = $this->timestampWithoutSignProvider();
+		$precisions = $this->simplePrecisionProvider();
+		foreach ( $timestamps as &$timestamp ) {
+			foreach ( $precisions as &$precision ) {
+				$this->auxTestGetHigherTimestamp( '+' . $timestamp, $precision );
+				$this->auxTestGetHigherTimestamp( '-' . $timestamp, $precision );
+			}
+		}
+		$timeValue = new TimeValue(
+			'+' . '2013-02-14' . 'T' . '12:51:02' . 'Z',
+			0, 0, 0,
+			TimeValue::PRECISION_MONTH,
+			TimeValue::CALENDAR_GREGORIAN
+		);
+		$timeValueCalculator = new TimeValueCalculator();
+		$timestampValue = $timeValueCalculator->getHigherTimestamp( $timeValue );
+		$this->assertEquals( 1362095999, $timestampValue );
+	}
+
+	/**
+	 * @param string $timestamp
+	 * @param int $precision
+	 */
+	private function auxTestGetLowerTimestamp( $timestamp, $precision ) {
+		$timeValueCalculator = new TimeValueCalculator();
+		$timeValue = new TimeValue(
+			$timestamp,
+			0, 0, 0,
+			$precision,
+			TimeValue::CALENDAR_GREGORIAN
+		);
+		$unixTimestampAsIs = $timeValueCalculator->getTimestamp( $timeValue );
+		$unixLowerTimestamp = $timeValueCalculator->getLowerTimestamp( $timeValue );
+		$this->assertGreaterThanOrEqual( $unixLowerTimestamp, $unixTimestampAsIs );
+
+		$timeValueBefore1 = new TimeValue(
+			$timestamp,
+			0, 1, 1,
+			$precision,
+			TimeValue::CALENDAR_GREGORIAN
+		);
+		$unixLowerTimestampBefore1 = $timeValueCalculator->getLowerTimestamp( $timeValueBefore1 );
+		$this->assertGreaterThan( $unixLowerTimestampBefore1, $unixLowerTimestamp );
+
+		$timeValueBefore2 = new TimeValue(
+			$timestamp,
+			0, 2, 2,
+			$precision,
+			TimeValue::CALENDAR_GREGORIAN
+		);
+		$unixLowerTimestampBefore2 = $timeValueCalculator->getLowerTimestamp( $timeValueBefore2 );
+		$this->assertGreaterThan( $unixLowerTimestampBefore2, $unixLowerTimestampBefore1 );
+	}
+
+	/**
+	 * @param string $timestamp
+	 * @param int $precision
+	 */
+	private function auxTestGetHigherTimestamp( $timestamp, $precision ) {
+		$timeValueCalculator = new TimeValueCalculator();
+		$timeValue = new TimeValue(
+			$timestamp,
+			0, 0, 0,
+			$precision,
+			TimeValue::CALENDAR_GREGORIAN
+		);
+		$unixTimestampAsIs = $timeValueCalculator->getTimestamp( $timeValue );
+		$unixHigherTimestamp = $timeValueCalculator->getHigherTimestamp( $timeValue );
+		$this->assertLessThanOrEqual( $unixHigherTimestamp, $unixTimestampAsIs );
+
+		$timeValueAfter1 = new TimeValue(
+			$timestamp,
+			0, 1, 1,
+			$precision,
+			TimeValue::CALENDAR_GREGORIAN
+		);
+		$unixHigherTimestampAfter1 = $timeValueCalculator->getHigherTimestamp( $timeValueAfter1 );
+		$this->assertLessThan( $unixHigherTimestampAfter1, $unixHigherTimestamp );
+
+		$timeValueAfter2 = new TimeValue(
+			$timestamp,
+			0, 2, 2,
+			$precision,
+			TimeValue::CALENDAR_GREGORIAN
+		);
+		$unixHigherTimestampAfter2 = $timeValueCalculator->getHigherTimestamp( $timeValueAfter2 );
+		$this->assertLessThan( $unixHigherTimestampAfter2, $unixHigherTimestampAfter1 );
```

Comment:
As far as I can see the vast majority of the tests relies on "greater/less than" calculations, but never does strict comparisons. I think this is a problem and hides a lot of possible issues. For example, a timestamp with a precision of a million years might as well give you a bunch of Unix timestamps that are all only a few seconds apart. No assertion here can ever catch this. :-(